### PR TITLE
Replace Reader/Accessor with a Mode enum

### PR DESCRIPTION
**Pull request by pinkwah**
Resolves: #6941 

This commit extends the "mode" concept and replaces the Reader/Accessor
pattern that was in `ert.storage` previously. This is to make it
possible to delete ensembles and experiments in a safe way.

The modes are `NONE` (`""`), `READ` (`"r"`) and `READ_WRITE` (`"w"`).
The -Reader classes are equivalent to `READ` and the -Accessor classes
are equivalent to `READ_WRITE`. `NONE` has no equivalent in the previous
code. With this change it is possible to downgrade the access-level of
objects without having to reopen them. This is in particular useful when
deleting ensembles and experiments.

Suppose there is a new method called `.delete_ensemble` which does what
it says. Consider the following code using the Reader/Accessor pattern:

```py
with open_storage(path, mode="w") as storage:
    ens = storage.get_ensemble_by_name("default")
    storage.delete_ensemble(ens)

    # ... sometime later:

    # ens is EnsembleAccessor, so writing is valid:
    ens.save_responses(group, real, data)
```

Uh-oh! We've accidentally saved data to a deleted object. Now, consider
with the capability pattern:

```py
with open_storage(path, mode="w") as storage:
    ens = storage.get_ensemble_by_name("default")
    storage.delete_ensemble(ens)

    # ... sometime later:

    # ens was reduced to Capability.NONE. Calling this method raises CapabilityError.
    ens.save_responses(group, real, data)
```

This should also make it easier to understand the module.

### CODE REVIEW

File: src/ert/storage/mode.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    def can_read(self) -> bool:
+        return self.__mode.can_read
+
+    @property
+    def can_write(self) -> bool:
+        return self.__mode.can_write
+
+    def assert_can_read(self) -> None:
+        if not self.can_read:
+            raise ModeError()
```
Comment (oyvindeide) [inline src/ert/storage/mode.py:53]: Never used it looks like?

File: tests/unit_tests/storage/test_local_storage.py [hunk 2]
```diff
@@ -128,37 +127,6 @@ def test_refresh(tmp_path):
...
-        assert isinstance(storage, local.LocalStorageReader)
-        assert not isinstance(storage, local.LocalStorageAccessor)
-
-    with open_storage(tmp_path, mode="r") as storage:
-        assert isinstance(storage, local.LocalStorageReader)
-        assert not isinstance(storage, local.LocalStorageAccessor)
-
-    with open_storage(tmp_path, mode="w") as storage:
-        assert isinstance(storage, local.LocalStorageReader)
-        assert isinstance(storage, local.LocalStorageAccessor)
```
Comment (oyvindeide) [inline tests/unit_tests/storage/test_local_storage.py:142]: Why are these removed?

File: src/ert/storage/mode.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ModeLiteral = Literal["", "r", "w"]
+
+
+class ModeError(ValueError):
+    pass
+
+
+class Mode(str, Enum):
+    NONE = ""
+    READ = "r"
```
Comment (oyvindeide) [inline src/ert/storage/mode.py:23]: No uses of this. Suggest either implementing it here, or remove this until the PR that does.

File: src/ert/gui/tools/run_analysis/run_analysis_tool.py [hunk 4]
```diff
@@ -177,11 +177,7 @@ def _enable_dialog(self, enable: bool):
             QApplication.setOverrideCursor(Qt.CursorShape.WaitCursor)
 
     def _init_analyse(self, source_fs: EnsembleReader, target: str):
-        with add_context_to_error(
```
Comment (oyvindeide) [inline src/ert/gui/tools/run_analysis/run_analysis_tool.py:180]: Why remove the context?

**Review by oyvindeide — COMMENTED** [review]
Looks like a simplification in my eyes!

File: tests/unit_tests/storage/test_local_storage.py (hunk 2 above)
Comment (pinkwah) [reply tests/unit_tests/storage/test_local_storage.py:142]: You get the same type regardless of mode, so the test makes no sense anymore.

File: src/ert/gui/tools/run_analysis/run_analysis_tool.py (hunk 4 above)
Comment (pinkwah) [reply src/ert/gui/tools/run_analysis/run_analysis_tool.py:180]: There is no more `to_accessor` function because this PR removes the Reader/Accessor distinction. This error can't be triggered anymore, so we don't need to add context to an error that can't occur.

File: src/ert/storage/mode.py (hunk 1 above)
Comment (pinkwah) [reply src/ert/storage/mode.py]: Removed

File: src/ert/storage/mode.py (hunk 3 above)
Comment (pinkwah) [reply src/ert/storage/mode.py:22]: Removed

File: tests/unit_tests/storage/test_local_storage.py (hunk 2 above)
Comment (oyvindeide) [reply tests/unit_tests/storage/test_local_storage.py:142]: But is there any tests of the input mode now? The test would have to be rewritten, could be something like:
```python
assert storage.mode == expected
```
